### ISSUE TITLE
Export and Inference YOLOv8 with FP16 Precision to TensorRT

### ISSUE THREAD

**Issue by pornpra** [ultralytics/ultralytics#7222]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi Ultralytics developer
I have some questions about how to export and inference YOLOv8 with FP16 precision. I used this code to export pretrained YOLOv8 with **batch size =1** and FP16 to TensorRT format without any error.

```
import torch
from ultralytics import YOLO

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

model_weight = 'yolov8s.pt'
model = YOLO(model_weight)
model.export(format='engine', device=device, half=True)
```
My questions are

1. After load the converted model (TensorRT format), Do I need to set parameter **half=True** inside Ultralytics's predict() function to enable FP16 precsion when inference or not ? 
2. How can I know the converted model is inferenced with FP16 not FP32 precision ?
3. How can I export YOLOv8 with FP16 precision and **batch size more than 1** to TensorRT format by using the latest Ultralytics version ?

My current enviroment:
Python: 3.8.10
Ultralytics: 8.0.230

Thanks

**Comment by glenn-jocher** [comment]
@pornpra hello! Great to hear you're working with YOLOv8 and exploring FP16 TensorRT export. Here are your answers:

1. You do not need to set `half=True` in the `predict()` function after exporting to TensorRT in FP16. The precision mode is already defined in your exported engine.
2. To confirm FP16 inference, you can benchmark the model using our Benchmark mode which profiles the speed and accuracy of the model - you'd typically notice increased speed with FP16.
3. To export YOLOv8 with FP16 precision and a batch size greater than 1, use the export function, specifying `batch_size` to your desired value greater than 1 along with `half=True`.

For more details on export and benchmarking specifics, please refer to our documentation. Have a great day! 😊

**Comment by pornpra** [comment]
@glenn-jocher Thanks for your answers.

Based on you recommendations, I use this command to export YOLOv8 with FP16 precision and batch size = 2 without any error:

`yolo export model='yolov8s.pt' format=engine device=0 batch=2 half=True`

After that, I compare inference results both yolov8s_b2_fp32.engine and yolov8s_b2_fp16.engine with 2 RTSPs using the **NVIDIA Jetson Orin Nano**. Both of two engines give me a bounding boxes when inference without any error. These are results:

yolov8s_b2_**fp32**.engine with 2 RTSPs
- engine file size: 46.2 MB
- speed: {'preprocess': 8.13138484954834, 'inference': 25.922298431396484, 'postprocess': 5.382299423217773}
- ram usage: 5323/6481MB (82.13%)

yolov8s_b2_**fp16**.engine with 2 RTSPs
- engine file size: 24.5 MB
- speed: {'preprocess': 6.451964378356934, 'inference': 11.924386024475098, 'postprocess': 4.981279373168945}
- ram usage: 5333/6481MB (82.28%)

Is this correct for YOLOv8 inference with FP16 precision and dynamic batch size ?

I'm not sure because the inference speed of yolov8s_b2_fp16.engine is half when compare with yolov8s_b2_fp32.engine but both of them consume the same level of ram usage. Inference with FP16 precision can reduce inference speed but can't reduce ram usage ?

**Comment by glenn-jocher** [comment]
@pornpra, your observations are correct. The inference speed improvement with the FP16 model (yolov8s_b2_fp16.engine) is expected due to the reduced precision, which allows for faster computation, especially on hardware that supports FP16 operations efficiently, like the NVIDIA Jetson Orin Nano.

Regarding RAM usage, FP16 precision primarily reduces the model's size on disk and the memory required for the model's weights during inference. However, the overall RAM usage includes additional factors such as the memory required for the input data, intermediate computations, and the framework's overhead, which may not significantly decrease with FP16. This explains why you see similar RAM usage for both FP32 and FP16 models.

In summary, your results indicate a successful FP16 inference with dynamic batch size, achieving a notable speedup without a reduction in RAM usage. Keep in mind that the benefits of FP16 may vary depending on the specific hardware and software environment.

**Comment by pornpra** [comment]
@glenn-jocher 

Thanks for your answers
I have one more question about TensorRT export with dynamic batch. Is the latest Ultralytics version supports export without specifying **batch = x** in the export command ?

`yolo export model=yolov8s.pt format=engine dynamic=True device=0 half=True`

**Comment by glenn-jocher** [comment]
@pornpra hello! Great to hear you're exploring FP16 precision with YOLOv8 and TensorRT. Here are the answers to your questions:

1. After exporting the model to TensorRT with `half=True`, you do not need to set `half=True` again in the `predict()` function. The exported model will already be optimized for FP16 inference.

2. To verify that the model is running with FP16 precision, you can check the TensorRT logs during the inference. The logs should indicate the precision mode the engine is using. Alternatively, you can use tools like NVIDIA's [Nsight Systems](https://developer.nvidia.com/nsight-systems) to profile the inference and see the precision details.

3. To export YOLOv8 with FP16 precision and a batch size greater than 1, you can specify the `batch` argument in the `export()` function. Here's an example for a batch size of 4:

   ```python
   model.export(format='engine', device=device, half=True, batch=4)
   ```

   Ensure that your GPU has enough memory to handle the larger batch size during inference.

Your environment setup looks good. If you encounter any issues or have further questions, feel free to reach out! 🚀
